File: analysis-service/app/models/contrarian_comeback.py
from typing import Literal, Optional

from pydantic import BaseModel, Field


class DailyBar(BaseModel):
    date: str
    high: Optional[float] = None
    low: Optional[float] = None
    close: Optional[float] = None
    volume: Optional[float] = None


class PriceTargetInfo(BaseModel):
    targetConsensus: Optional[float] = None
    targetHigh: Optional[float] = None
    targetLow: Optional[float] = None


class GradeRecord(BaseModel):
    gradingCompany: Optional[str] = None
    newGrade: Optional[str] = None
    action: Optional[str] = None  # 'upgrade' | 'downgrade' | other, as reported by FMP
    date: Optional[str] = None


class InsiderTrade(BaseModel):
    transactionDate: Optional[str] = None
    transactionType: Optional[str] = None
    acquisitionOrDisposition: Optional[str] = None
    securitiesTransacted: Optional[float] = None
    price: Optional[float] = None
    reportingName: Optional[str] = None


class NewsItem(BaseModel):
    date: Optional[str] = None
    title: str
    source: Optional[str] = None
    url: Optional[str] = None


class IncomeStatementPeriod(BaseModel):
    revenue: Optional[float] = None
    grossProfit: Optional[float] = None


class ContrarianComebackData(BaseModel):
    """Everything Node fetches/normalizes from FMP (+ optional Finnhub news)
    for one symbol. Shared by both the gate-preview and submit requests —
    both endpoints re-derive the same auto-checks from this, independently
    and statelessly, matching Contrarian Finder's per-batch
    assembleUniverse() philosophy (safe/cheap to recompute)."""

    symbol: str
    companyName: Optional[str] = None
    sector: Optional[str] = None
    exchange: Optional[str] = None
    price: float
    marketCap: Optional[float] = None
    yearHigh: Optional[float] = None
    peRatio: Optional[float] = None  # profile.pe ?? quote.pe, Node-normalized
    incomeStatements: list[IncomeStatementPeriod] = []  # 2 most recent annual periods, newest first
    dailyBars: list[DailyBar] = []  # stock, newest-first, up to 1000 bars
    etfSymbol: Optional[str] = None  # SECTOR_ETF[sector], resolved by Node
    etfDailyBars: list[DailyBar] = []  # newest-first, up to 260 bars; empty if no ETF mapped
    priceTarget: Optional[PriceTargetInfo] = None
    grades: list[GradeRecord] = []
    insiderTrades: list[InsiderTrade] = []
    news: list[NewsItem] = []
    # Phase 2 - most recent annual balance-sheet / cash-flow figures only (no
    # prior-year comparison needed for DE/currentR/FCF/cashRun, unlike revenue growth).
    totalDebt: Optional[float] = None
    totalStockholdersEquity: Optional[float] = None
    totalCurrentAssets: Optional[float] = None
    totalCurrentLiabilities: Optional[float] = None
    cashAndCashEquivalents: Optional[float] = None
    operatingCashFlow: Optional[float] = None
    capitalExpenditure: Optional[float] = None


class ContrarianComebackGateResponse(BaseModel):
    symbol: str
    check1Pass: bool
    drawdownPct: float
    dd52w: float
    dd4y: float
    check3Status: Literal['pass', 'override_available', 'hard_block']
    etfSymbol: Optional[str] = None
    etfReturn6M: Optional[float] = None
    check4Pass: bool
    recentNews: list[NewsItem] = []
    insiderSignal: str
    insiderBuys: int
    insiderSells: int
    analystUpgrades90d: int
    analystDowngrades90d: int
    priceTargetAvg: Optional[float] = None
    analystUpsidePct: Optional[float] = None
    # Non-null only when an auto-check (1, 3-hard-block, or 4) already fails -
    # the frontend can skip rendering the checkbox form entirely and show the
    # rejection immediately, matching the source app's init() short-circuit.
    failedCheck: Optional[str] = None
    reason: Optional[str] = None
    route: Optional[str] = None


class ContrarianComebackSubmitRequest(ContrarianComebackData):
    # min_length=1 rejects an empty selection with a 422 here too, in
    # addition to the Node controller's own 400 check on the same field -
    # defense in depth, not the primary validation layer (that's Node's job).
    breakdownTypes: list[str] = Field(min_length=1)
    catalystAnswer: Literal['yes', 'no']
    check3Override: bool = False
    check3OverrideReason: Optional[str] = None


class ScoreBreakdown(BaseModel):
    breakdown: int
    sector: int
    technical: int
    value: int
    catalyst: int
    total: int
    verdict: Literal['HIGH', 'MODERATE', 'SPECULATIVE', 'AVOID']
    hybridCapActive: bool
    sectorOverrideCapActive: bool


class WeeklyTechnicals(BaseModel):
    weeklyRsi: Optional[float] = None
    obvTrend: Literal['up', 'down', 'flat', 'insufficient_data']
    volumeDrying: bool
    sma200w: Optional[float] = None


class FibonacciLevels(BaseModel):
    swingLow: float
    athPrice: float
    fib382: float
    fib618: float
    fib100: float


class FundamentalMetric(BaseModel):
    value: Optional[float] = None
    tier: Optional[Literal['green', 'yellow', 'red']] = None


class FundamentalHealth(BaseModel):
    debtToEquity: FundamentalMetric
    currentRatio: FundamentalMetric
    freeCashFlow: FundamentalMetric
    revenueGrowthPct: FundamentalMetric
    grossMarginPct: FundamentalMetric
    cashRunwayMonths: FundamentalMetric
    positiveFcf: bool  # true -> frontend shows "Positive FCF" instead of a runway number


class CatalystPipeline(BaseModel):
    recentInsiderTrades: list[InsiderTrade] = []
    recentGrades: list[GradeRecord] = []
    news: list[NewsItem] = []


class TrancheEntry(BaseModel):
    label: str
    sizePct: int
    priceLow: float
    priceHigh: float
    trigger: str


class StagedEntry(BaseModel):
    tranches: list[TrancheEntry]
    hardStop: float
    capLabel: Literal['Large-Cap', 'Mid-Cap', 'Small-Cap']
    # Mirrors the source app's isMidCap flag verbatim - true for anything
    # under $10B (small-caps included), not exclusively "true mid-cap". The
    # liquidity note it gates applies equally to small-caps, so this isn't a
    # naming bug to "fix".
    isMidCap: bool


class RecoveryTarget(BaseModel):
    label: str
    horizon: str
    price: float
    returnPct: float


class AnalystConsensusTarget(BaseModel):
    low: float
    high: float
    average: float
    returnPct: float


class RecoveryTargets(BaseModel):
    conservative: RecoveryTarget
    baseCase: RecoveryTarget
    bullCase: RecoveryTarget
    analystConsensus: Optional[AnalystConsensusTarget] = None
    riskRewardRatio: Optional[float] = None


class ContrarianComebackSubmitResponse(BaseModel):
    symbol: str
    format: Literal['A', 'B']
    # Format B
    failedCheck: Optional[str] = None
    reason: Optional[str] = None
    route: Optional[str] = None
    # Format A
    companyName: Optional[str] = None
    sector: Optional[str] = None
    exchange: Optional[str] = None
    price: Optional[float] = None
    drawdownPct: Optional[float] = None
    breakdownTypes: list[str] = []
    hybridCap: bool = False
    check3Override: bool = False
    check3OverrideReason: Optional[str] = None
    etfSymbol: Optional[str] = None
    etfReturn6M: Optional[float] = None
    score: Optional[ScoreBreakdown] = None
    technicals: Optional[WeeklyTechnicals] = None
    fibonacci: Optional[FibonacciLevels] = None
    fundamentalHealth: Optional[FundamentalHealth] = None
    catalystPipeline: Optional[CatalystPipeline] = None
    stagedEntry: Optional[StagedEntry] = None
    recoveryTargets: Optional[RecoveryTargets] = None

File: frontend/src/api/contrarianComeback.ts
import { useMutation } from '@tanstack/react-query';
import { apiFetch } from './client';

// Mirrors analysis-service/app/models/contrarian_comeback.py field-for-field
// - no shared-schema codegen in this repo, keep in sync by hand if either
// shape changes (same caveat as backend/src/services/analysisService.ts's
// own copy).

export interface NewsItem {
  date: string | null;
  title: string;
  source: string | null;
  url: string | null;
}

export interface ContrarianComebackGateResult {
  symbol: string;
  check1Pass: boolean;
  drawdownPct: number;
  dd52w: number;
  dd4y: number;
  check3Status: 'pass' | 'override_available' | 'hard_block';
  etfSymbol: string | null;
  etfReturn6M: number | null;
  check4Pass: boolean;
  recentNews: NewsItem[];
  insiderSignal: string;
  insiderBuys: number;
  insiderSells: number;
  analystUpgrades90d: number;
  analystDowngrades90d: number;
  priceTargetAvg: number | null;
  analystUpsidePct: number | null;
  failedCheck: string | null;
  reason: string | null;
  route: string | null;
}

export interface ScoreBreakdown {
  breakdown: number;
  sector: number;
  technical: number;
  value: number;
  catalyst: number;
  total: number;
  verdict: 'HIGH' | 'MODERATE' | 'SPECULATIVE' | 'AVOID';
  hybridCapActive: boolean;
  sectorOverrideCapActive: boolean;
}

export interface WeeklyTechnicals {
  weeklyRsi: number | null;
  obvTrend: 'up' | 'down' | 'flat' | 'insufficient_data';
  volumeDrying: boolean;
  sma200w: number | null;
}

export interface FibonacciLevels {
  swingLow: number;
  athPrice: number;
  fib382: number;
  fib618: number;
  fib100: number;
}

export interface InsiderTrade {
  transactionDate: string | null;
  transactionType: string | null;
  acquisitionOrDisposition: string | null;
  securitiesTransacted: number | null;
  price: number | null;
  reportingName: string | null;
}

export interface GradeRecord {
  gradingCompany: string | null;
  newGrade: string | null;
  action: string | null;
  date: string | null;
}

export interface FundamentalMetric {
  value: number | null;
  tier: 'green' | 'yellow' | 'red' | null;
}

export interface FundamentalHealth {
  debtToEquity: FundamentalMetric;
  currentRatio: FundamentalMetric;
  freeCashFlow: FundamentalMetric;
  revenueGrowthPct: FundamentalMetric;
  grossMarginPct: FundamentalMetric;
  cashRunwayMonths: FundamentalMetric;
  positiveFcf: boolean;
}

export interface CatalystPipeline {
  recentInsiderTrades: InsiderTrade[];
  recentGrades: GradeRecord[];
  news: NewsItem[];
}

export interface TrancheEntry {
  label: string;
  sizePct: number;
  priceLow: number;
  priceHigh: number;
  trigger: string;
}

export interface StagedEntry {
  tranches: TrancheEntry[];
  hardStop: number;
  capLabel: 'Large-Cap' | 'Mid-Cap' | 'Small-Cap';
  isMidCap: boolean;
}

export interface RecoveryTarget {
  label: string;
  horizon: string;
  price: number;
  returnPct: number;
}

export interface AnalystConsensusTarget {
  low: number;
  high: number;
  average: number;
  returnPct: number;
}

export interface RecoveryTargets {
  conservative: RecoveryTarget;
  baseCase: RecoveryTarget;
  bullCase: RecoveryTarget;
  analystConsensus: AnalystConsensusTarget | null;
  riskRewardRatio: number | null;
}

export interface ContrarianComebackSubmitResult {
  symbol: string;
  format: 'A' | 'B';
  failedCheck: string | null;
  reason: string | null;
  route: string | null;
  companyName: string | null;
  sector: string | null;
  exchange: string | null;
  price: number | null;
  drawdownPct: number | null;
  breakdownTypes: string[];
  hybridCap: boolean;
  check3Override: boolean;
  check3OverrideReason: string | null;
  etfSymbol: string | null;
  etfReturn6M: number | null;
  score: ScoreBreakdown | null;
  technicals: WeeklyTechnicals | null;
  fibonacci: FibonacciLevels | null;
  fundamentalHealth: FundamentalHealth | null;
  catalystPipeline: CatalystPipeline | null;
  stagedEntry: StagedEntry | null;
  recoveryTargets: RecoveryTargets | null;
}

export interface ContrarianComebackSubmitInput {
  symbol: string;
  breakdownTypes: string[];
  catalystAnswer: 'yes' | 'no';
  check3Override?: boolean;
  check3OverrideReason?: string;
}

// A mutation, not a query - same reasoning as useLongTermAnalysis: both are
// triggered on-demand by a ticker lookup, not something to auto-refetch/
// cache by symbol.
export function useContrarianComebackGate() {
  return useMutation({
    mutationFn: (symbol: string) =>
      apiFetch<ContrarianComebackGateResult>(`/analysis/contrarian-comeback/${encodeURIComponent(symbol)}/gate`),
  });
}

export function useContrarianComebackSubmit() {
  return useMutation({
    mutationFn: ({ symbol, ...body }: ContrarianComebackSubmitInput) =>
      apiFetch<ContrarianComebackSubmitResult>(`/analysis/contrarian-comeback/${encodeURIComponent(symbol)}`, {
        method: 'POST',
        body: JSON.stringify(body),
      }),
  });
}
